### DISCUSSION TITLE
Cannot generate pdf on prod server ubuntu 22

### DISCUSSION THREAD

**Discussion by aeq-dev** [spatie/browsershot#736] [General]
Hi, 
I cannot generate pdf on production server ubuntu 22, while I'm able to generate on local  !
Here's my code : 

```
$invoice = Invoice::find($id);
        $html = view(
            'accounting::invoices.pdf',
            [
                'invoice' => $invoice,                
            ]
        )->render();

        $path = storage_path("/pdfs/");
        Browsershot::html($html)
            ->setChromePath(env('PDF_CHROME_PATH', '/snap/bin/chromium'))
            ->setNodeBinary(env('PDF_NODE_BINARY', '/usr/bin/node'))
            ->setNpmBinary(env('PDF_NPM_BINARY', '/usr/bin/npm'))
            ->waitUntilNetworkIdle()
            ->noSandbox()
            ->ignoreHttpsErrors()
            ->timeout(500)
            ->format('A4')
            ->setCustomTempPath(storage_path('pdfs'))
            ->savePdf($path . $invoice->number . '.pdf');

```
Here's the error message : 
```
Symfony \ Component \ Process \ Exception \ ProcessFailedException
PHP 8.1.2-1ubuntu2.11  10.11.0

The command "PATH=$PATH:/usr/local/bin:/opt/homebrew/bin NODE_PATH=`/usr/bin/node /usr/bin/npm root -g` /usr/bin/node '/var/www/dev/releases/140/vendor/spatie/browsershot/src/../bin/browser.js' '{"url":"file:\/\/\/var\/www\/dev\/releases\/140\/storage\/pdfs\/1983874840-0676282001685054994\/index.html","action":"pdf","options":{"path":"\/var\/www\/dev\/releases\/140\/storage\/pdfs\/INV-2023-001.pdf","args":["--no-sandbox"],"viewport":{"width":800,"height":600},"displayHeaderFooter":false,"executablePath":"\/snap\/bin\/chromium","waitUntil":"networkidle0","ignoreHttpsErrors":true,"timeout":500000,"format":"A4"}}'" failed. 
Exit Code: 1(General error) Working directory: /var/www/dev/releases/140/public Output: ================ Error Output: 
================ [1m[43m[30m Puppeteer old Headless deprecation warning:[0m[33m In the near feature `headless: true` will default to the new Headless mode for Chrome instead of the old Headless implementation. For more information, please see https://developer.chrome.com/articles/new-headless/. Consider opting in early by passing `headless: "new"` to `puppeteer.launch()` If you encounter any bugs, please report them to https://github.com/puppeteer/puppeteer/issues/new/choose.[0m 
Error: Failed to launch the browser process! cmd_run.go:1046: WARNING: 
cannot create user data directory: cannot create snap home dir: mkdir /var/www/snap: permission denied /system.slice/php8.1-fpm.service is not a snap cgroup 
TROUBLESHOOTING: https://pptr.dev/troubleshooting at Interface.onClose (/usr/lib/node_modules/puppeteer/node_modules/@puppeteer/browsers/lib/cjs/launch.js:262:24) at Interface.emit (node:events:525:35) at Interface.close 
(node:internal/readline/interface:533:10) at Socket.onend (node:internal/readline/interface:259:10) at Socket.emit (node:events:525:35) at endReadableNT (node:internal/streams/readable:1359:12) at process.processTicksAndRejections (node:internal/process/task_queues:82:21) 
```
Thanks for helping me 
NB: I can generate it using tinker :/

**Comment by jameshulse** [comment]
```
/var/www/snap: permission denied /system.slice/php8.1-fpm.service is not a snap cgroup
```

This is your error. I think it basically means that you have installed Chrome via [Snap](https://ubuntu.com/core/services/guide/snaps-intro), but your PHP web user (php8.1-fpm.service) doesn't have permission to use snaps.

You could either add the user to the group, or install Chrome via another method.

**Reply by jr-cobweb** [reply]
Hey @jameshulse . I'm having the same issue. I don't suppose you could expand on adding the user to the group. Is there a snap command for that, or are you talking about standard/OS level groups?

**Reply by talelmishali** [reply]
Any update on that by any chance? Thanks 🙏🏻

**Reply by liamseys** [reply]
> ```
> /var/www/snap: permission denied /system.slice/php8.1-fpm.service is not a snap cgroup
> ```

I have this issue as well, is there a way to fix this?

**Comment by nepa-anujpandey** [comment]
@aeq-dev Did you manage to solve this issue? Plus how did you manage the permission issue?

This is the error I'm getting.
```
Error Output:
================
sh: 1: npm: not found
sh: 1: /root/.nvm/versions/node/v22.2.0/bin/node: Permission denied
```

**Comment by alexivashchenko** [comment]
I assume you can run it with Tinker since Tinker does not use NGINX and php-fpm which does not belong to snap cgroup and has no permissions to run Chrome that was installed under root user and snap.

To make it work, you need 2 things:
1. To have a real www-data user (with home directory) (`sudo mkdir /home/www-data`, `sudo usermod -d /home/www-data www-data -s /bin/bash`).
2. To have chrome browser installed by www-data user (Puppeteer has a command for this: `sudo -u www-data npx puppeteer browsers install chrome`).

Yesterday I managed to make Browsershoot work with next environment: Ubuntu 24.04LST, php8.3, NGINX, Node (v18.19.1), Spatie/Browsershot (Puppeteer, Chrome 128.0.6613.119). So it should be possible.

Here are detailed instructions: https://stackoverflow.com/a/78956369/3163536

**Comment by slakbal** [comment]
@alexivashchenko do you maybe know how this can be achieved with a forge provisioned server running on ubuntu and Arm64 architecture? Since forge configure nginx I assume one can't tamper too much with it.

22.04.4 LTS (GNU/Linux 6.5.0-1027-oracle aarch64)

I tried the below:

```bash
bash sudo chown -R forge:forge /snap/bin/chromium
```

It works, but after a while I get the following error:

Error Output:
================
Error: Failed to launch the browser process!
/system.slice/php8.2-fpm.service is not a snap cgroup


TROUBLESHOOTING: https://pptr.dev/troubleshooting

    at Interface.onClose (/home/forge/tools.kayagroup.co.za/node_modules/@puppeteer/browsers/lib/cjs/launch.js:312:24)
    at Interface.emit (node:events:526:35)
    at Interface.close (node:internal/readline/interface:533:10)
    at Socket.onend (node:internal/readline/interface:259:10)
    at Socket.emit (node:events:526:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)

**Reply by alexivashchenko** [reply]
I guess, if there is `forge` user who probably a big boss there, I would try to install chrome by `forge` user like: `sudo -u forge npx puppeteer browsers install chrome`
Sorry, cannot say for sure since I have no experience with Forge.

**Reply by slakbal** [reply]
Appreciate the feedback... will give it a shot

**Reply by DezineHQ** [reply]
Hi . I installed Puppeteer globally and it works for a few hours but then I get the following error.
/system.slice/php8.3-fpm.service is not a snap cgroup

**Reply by slakbal** [reply]
yip me too @DezineHQ , any solution? I've been breaking the internet to find a solution. Weirdest thing, as soon as I simply ssh into the server it works, and as soon as I close the terminal session it stops working again.

**Reply by johnrich85** [reply]
Only fix I found was to uninstall the snap version and install from linux mint repos instead

**Reply by striderwhite** [reply]
@johnrich85 @DezineHQ could either of you explain this workaround? Any time I try to install chromium it falls back to using snap.

**Reply by slakbal** [reply]
Yeah this is a real pain ...

**Reply by striderwhite** [reply]
@slakbal I found a workaround, just install google chrome directly (without a package manager) and use that. Here are some instructions

```bash
# download chrome package
wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb

# try installing it, if it gives errors, install and missing deps, I had to install fonts-liberation after running this:
sudo dpkg -i google-chrome-stable_current_amd64.deb

# for me I had to install this missing dep
sudo apt-get install fonts-liberation

# install (again) if needed
sudo dpkg -i google-chrome-stable_current_amd64.deb

# remove old file
rm -rf google-chrome-stable_current_amd64.deb 

# check if it installed 
google-chrome-stable --version

# now find the path, use this when pointing browsershot/whatever
whereis google-chrome 
# probably is /usr/bin/google-chrome
# EX:
# $browsershot->setChromePath('/usr/bin/google-chrome');
```

**Reply by slakbal** [reply]
Hey @striderwhite really appreciate your help. This is such pain. I have tried the steps and still get an error. This is a clean ARM 64 server, provisioned by forge. I've ssh'ed into the machine and tried the steps but get an architecture error.

![Screenshot 2024-12-03 at 11 55 01](https://github.com/user-attachments/assets/9be55519-f832-4237-aee0-c7131d6bdb7d)

Again thanks for your help, but looks like I need an ARM64 version of the .deb, which I think does not exist yet.

**Reply by ghost** [reply]
@slakbal Did you find a stable solution for arm servers yet?

**Reply by slakbal** [reply]
Hey @maxbaumhoegger-cnnctvstn nope not really...

**Reply by striderwhite** [reply]
@maxbaumhoegger-cnnctvstn maybe you could try using chromium instead of google chrome, there might be an arm build for it.

**Comment by DezineHQ** [comment]
I removed the Snap version and installed as per link below. Not ideal but it worked.How to install Chromium without snap?askubuntu.comDezine HQOn 26 Sep 2024, at 17:10, John Richardson ***@***.***> wrote:﻿
Only fix I found was to uninstall the snap version and install from linux mint repos instead

**Comment by WesleyMAraujo** [comment]
I managed to solve the problem in the following way

`$chrome = new Chrome('/snap/bin/chromium');`

when I was having this problem it was in the following way

`$chrome = new Chrome(null, '/snap/bin/chromium');`

**Reply by slakbal** [reply]
Did your setup work with the Spatie Browsershot package?

```
Browsershot::html($html)
            ->setChromePath(env('PDF_CHROME_PATH', '/snap/bin/chromium'))
            ->setNodeBinary(env('PDF_NODE_BINARY', '/usr/bin/node'))
            ->setNpmBinary(env('PDF_NPM_BINARY', '/usr/bin/npm'))
            ->waitUntilNetworkIdle()
            ->noSandbox()
            ->ignoreHttpsErrors()
            ->timeout(500)
            ->format('A4')
            ->setCustomTempPath(storage_path('pdfs'))
            ->savePdf($path . $invoice->number . '.pdf');
```

**Comment by federicosolfa** [comment]
On systems using systemd, Snap applications (like Chromium) run inside the user’s `systemd --user` session. If the user (forge in this case) is not logged in, systemd stops the user session, so Snap apps can’t run and you may get errors like:

_**php8.1-fpm.service is not a snap cgroup**_

Enable linger for the forge user to keep their systemd user session active even when they are not logged in:

`sudo loginctl enable-linger forge`

This prevents the error and ensures Snap and Chromium work correctly at all times.

**Comment by boris-glumpler** [comment]
Here's how I solved this issue on Ubuntu 24.04 (arm):

1. Add the `xtradeb/apps` ppa
2. Install their`chromium` package, which is an apt package and not a snap one
3. Tell puppeteer about it: `$browsershot->setChromePath('/usr/bin/chromium')`
4. Install any missing dependencies that puppeteer complains about in `laravel.log`
5. Success!
